authorization-common/src/test/java/com/sequenceiq/authorization/EnforceAuthorizationLogicsUtil.java

```diff
@@ -62,18 +124,239 @@ public static void testIfControllerClassHasAuthorizationAnnotation() {
+                    }
+                    if (actualTypeArguments.length == 1 && generic.genericType.isAssignableFrom((Class<?>) actualTypeArguments[0])) {
+                        count++;
+                    }
+                }
+            }
+            String errorMessage = "Return " + method.getReturnType().getSimpleName() + " type is not one of" +
+                    Arrays.stream(genericTypes)
+                            .map(Generic::toString)
+                            .collect(Collectors.joining(",", "[", "]"));
+            if (count != 1) {
+                return Optional.of(invalid(method, errorMessage));
+            } else {
+                return Optional.empty();
+            }
+        };
+    }
+
+    private static Function<Pair<Method, Class<?>>, Optional<String>> resourceObject() {
+        return ctx -> {
+            Method method = ctx.getKey();
+            Class<?> type = ctx.getValue();
+            boolean hasValidField = false;
+            for (Field field : FieldUtils.getFieldsWithAnnotation(type, ResourceObjectField.class)) {
+                ResourceObjectField annotation = field.getAnnotation(ResourceObjectField.class);
+                if (Set.of(CRN, NAME).contains(annotation.variableType())) {
+                    if (String.class.equals(field.getType())) {
+                        hasValidField = true;
+                    } else {
+                        return Optional.of(invalid(method, field.getName()
+                                + " @" + ResourceObjectField.class + " must be a String in " + type.getSimpleName()));
+                    }
+                } else if (Set.of(CRN_LIST, NAME_LIST).contains(annotation.variableType())) {
+                    if (isOneOf(field, list(String.class), set(String.class))) {
+                        hasValidField = true;
+                    } else {
+                        return Optional.of(invalid(method, field.getName()
+                                + " @" + ResourceObjectField.class + " must be a Set<String> or List<String> in " + type.getSimpleName()));
+                    }
+                }
+            }
+            if (hasValidField) {
+                return Optional.empty();
+            } else {
+                return Optional.of(invalid(method, "Missing @" + ResourceObjectField.class + " annotaion in " + type.getSimpleName()));
+            }
+        };
+    }
+
+    private static boolean isOneOf(Field field, Generic... genericTypes) {
+        for (Generic generic : genericTypes) {
+            if (generic.wrapperType.isAssignableFrom(field.getType())) {
+                Type[] actualTypeArguments = ((ParameterizedType) field.getGenericType()).getActualTypeArguments();
+                if (actualTypeArguments.length == 1 && actualTypeArguments[0].equals(generic.genericType)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public static class Generic {
```

Comment:
can we extract this into different class with more specific name?

Reply:
yep